"Self-filtering may fail" because of missing waitForTransform

I keep sporadically getting errors like this when using the DepthImageOctomapUpdater:

```
Transform error: Lookup would require extrapolation into the future.  Requested time 1401446451.006236446 but the latest data is at time 1401446450.937553891, when looking up transform from frame [katana_motor2_lift_link] to frame [kinect_rgb_optical_frame]
[ERROR] [1401446451.007936297]: Transform cache was not updated. Self-filtering may fail.
```

This message originates from [depth_image_octomap_updater.cpp:285](https://github.com/ros-planning/moveit_ros/blob/hydro-devel/perception/depth_image_octomap_updater/src/depth_image_octomap_updater.cpp#L285). It looks to me as if there was a `waitForTransform()` missing in [planning_scene_monitor.cpp:818](https://github.com/ros-planning/moveit_ros/blob/56c1b9465fd86795baaff3fb94746e57aeb09ddb/planning/planning_scene_monitor/src/planning_scene_monitor.cpp#L818): When `lookupTransform()` is called with the time stamp from the the DepthImage, sometimes the TF buffer for that time isn't filled yet (in this case, it is 0.07 seconds behind), thereby triggering a race condition.

The Katana arm I'm working with can only publish joint states at 25 Hz, so maybe that's why this race condition happens more frequently than with other robots.

--- a/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h
+++ b/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h
@@ -493,6 +493,11 @@ private:
   // This field is protected by state_pending_mutex_
   ros::WallDuration dt_state_update_;
 
+  /// the amount of time to wait when looking up transforms
+  // Setting this to a non-zero value resolves issues when the sensor data is
+  // arriving so fast that it is preceding the transform state.
+  ros::Duration shape_transform_cache_lookup_wait_time_;
+
   /// timer for state updates.
   // Check if last_state_update_ is true and if so call updateSceneWithCurrentState()
   // Not safe to access from callback functions.

--- a/planning/planning_scene_monitor/src/planning_scene_monitor.cpp
+++ b/planning/planning_scene_monitor/src/planning_scene_monitor.cpp
@@ -216,6 +216,14 @@ void planning_scene_monitor::PlanningSceneMonitor::initialize(const planning_sce
   last_update_time_ = ros::Time::now();
   last_state_update_ = ros::WallTime::now();
   dt_state_update_ = ros::WallDuration(0.1);
+
+  double temp_wait_time;
+  nh_.param(
+      robot_description_ + "_planning/shape_transform_cache_lookup_wait_time",
+      temp_wait_time,
+      0.05);
+  shape_transform_cache_lookup_wait_time_ = ros::Duration(temp_wait_time);
+
   state_update_pending_ = false;
   state_update_timer_ = nh_.createWallTimer(dt_state_update_,
                                             &PlanningSceneMonitor::stateUpdateTimerCallback,
@@ -846,6 +854,7 @@ bool planning_scene_monitor::PlanningSceneMonitor::getShapeTransformCache(const 
     for (LinkShapeHandles::const_iterator it = link_shape_handles_.begin() ; it != link_shape_handles_.end() ; ++it)
     {
       tf::StampedTransform tr;
+      tf_->waitForTransform(target_frame, it->first->getName(), target_time, shape_transform_cache_lookup_wait_time_);
       tf_->lookupTransform(target_frame, it->first->getName(), target_time, tr);
       Eigen::Affine3d ttr;
       tf::transformTFToEigen(tr, ttr);
@@ -855,6 +864,7 @@ bool planning_scene_monitor::PlanningSceneMonitor::getShapeTransformCache(const 
     for (AttachedBodyShapeHandles::const_iterator it = attached_body_shape_handles_.begin() ; it != attached_body_shape_handles_.end() ; ++it)
     {
       tf::StampedTransform tr;
+      tf_->waitForTransform(target_frame, it->first->getAttachedLinkName(), target_time, shape_transform_cache_lookup_wait_time_);
       tf_->lookupTransform(target_frame, it->first->getAttachedLinkName(), target_time, tr);
       Eigen::Affine3d transform;
       tf::transformTFToEigen(tr, transform);
@@ -863,6 +873,7 @@ bool planning_scene_monitor::PlanningSceneMonitor::getShapeTransformCache(const 
     }
     {
       tf::StampedTransform tr;
+      tf_->waitForTransform(target_frame, scene_->getPlanningFrame(), target_time, shape_transform_cache_lookup_wait_time_);
       tf_->lookupTransform(target_frame, scene_->getPlanningFrame(), target_time, tr);
       Eigen::Affine3d transform;
       tf::transformTFToEigen(tr, transform);

